### PR TITLE
Make location blue circle radius configurable

### PR DESCRIPTION
**Pull request by vincent-karuri**
Partial fix for https://github.com/OpenSRP/opensrp-client-reveal/issues/40.

Related to https://github.com/onaio/kujaku/issues/38

### CODE REVIEW

File: library/build.gradle [hunk 1]
```diff
@@ -21,7 +21,7 @@ apply plugin: 'realm-android'
 apply plugin: 'com.jfrog.bintray'
 apply plugin: 'maven-publish'
 
-version '0.6.1'
+version '0.6.2'
```
Comment (ekigamba) [inline library/build.gradle]: I don't think that this should be here? Or rather should have been changed in this PR

File: library/src/main/java/io/ona/kujaku/exceptions/LocationComponentInitializationException.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package io.ona.kujaku.exceptions;
+
+/**
+ * @author Vincent Karuri
+ */
+public class LocationComponentInitializationException extends RuntimeException {
```
Comment (ekigamba) [inline library/src/main/java/io/ona/kujaku/exceptions/LocationComponentInitializationException.java]: IMO, this tends to be a bit dangerous unless really crucial that the whole application crashes while throwing this exception. Would you agree with this article https://docs.oracle.com/javase/tutorial/essential/exceptions/runtime.html?

File: library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java [hunk 3]
```diff
@@ -70,11 +70,22 @@
      * intervention. If the MY LOCATION BUTTON is visible, it will turn blue as long as this mode is on.
      * This can be turned off by the user if s/he touches the map to scroll it to a specific location.
      *
-     * @param focusOnMyLocation
+     * @param focusOnMyLocation Whether to focus on the user's current location or not
      */
     void focusOnUserLocation(boolean focusOnMyLocation);
```
Comment (ekigamba) [inline library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: I tried testing out the other activities with this functionality, it seems that the functionality has broken and I have to click on the **MY LOCATION BUTTON** multiple times(like 10-20) for the map to zoom to my location as it did previously. 
1. Is this the expected functionality or has is it a bug?
2. If this is the expected functionality, why change this to the current implementation ? In my humble opinion, it seems a bit more difficult and takes more effort for a user to actually get to see their current location

Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: You're right, it's harder now.

It's not a bug, just a CameraMode setting for the LocationComponent.

Easy fix.

Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: Is this configurable or is this the default for the feature currently? If it's not configurable and you  need it to be, could you do that and have the default as it was before.

Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: One can always access the LocationComponent and reset this, I have removed this behavior as the default setting.

File: library/src/main/java/io/ona/kujaku/exceptions/LocationComponentInitializationException.java (hunk 2 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/exceptions/LocationComponentInitializationException.java]: I had thought about this. My reasoning is:

```
One of the most important features of the GeoWidget is providing location services to the host app.

If the LocationComponent is wrongly initialized, it will affect the experience on the whole app and make it unusable for this purpose.

Basically as bad as performing operations on a null value, don't you think?
```

File: library/build.gradle (hunk 1 above)
Comment (vincent-karuri) [reply library/build.gradle]: Not sure what the suggestiion is here. 

When/where should this be done/placed?

Comment (ekigamba) [reply library/build.gradle]: Revert it back to the previous version

File: library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java (hunk 3 above)
Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: 1. It's a bit jerky. I don't know why?
2. The buffer radius and it’s blue appearance(of the location component) disappears after some time, is this going to be a problem for the specific use cases of Reveal and other host apps
screenshot_20190220-115336

File: library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    public static MapboxLocationComponentWrapper getInstance() {
+        if (mapboxLocationComponentWrapper == null) {
+            throw new LocationComponentInitializationException("The MapboxLocationComponentWrapper has not been initialized! " +
+                    "Please initialize it by calling init before calling the getInstance() method.");
+        }
+        return mapboxLocationComponentWrapper;
+    }
+
+    public LocationComponent getLocationComponent() {
+       return MapboxLocationComponentWrapper.locationComponent;
```
Comment (githengi) [inline library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: There is no need to qualify `MapboxLocationComponentWrapper.locationComponent` locationComponent is adequate

File: library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    private static LocationComponent locationComponent;
+
+    private static MapboxLocationComponentWrapper mapboxLocationComponentWrapper;
+
+    private MapboxLocationComponentWrapper() {}
+
+    @SuppressWarnings( {"MissingPermission"})
+    public static void init(@NonNull MapboxMap mapboxMap,@NonNull Context context) {
+        mapboxLocationComponentWrapper = new MapboxLocationComponentWrapper();
+        LocationComponent locationComponent = mapboxMap.getLocationComponent();
```
Comment (githengi) [inline library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: Remove the definition of this private variable and assign the static variable `locationComponent`

File: library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import com.mapbox.mapboxsdk.location.modes.CameraMode;
+import com.mapbox.mapboxsdk.location.modes.RenderMode;
+import com.mapbox.mapboxsdk.maps.MapboxMap;
+
+import io.ona.kujaku.exceptions.LocationComponentInitializationException;
+
+/**
+ * @author Vincent Karuri
+ */
+public class MapboxLocationComponentWrapper {
```
Comment (githengi) [inline library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: Could this class be moved to a function inside KukakuMapView as Its only initializing the location component.

File: library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+    private static LocationComponent locationComponent;
```
Comment (githengi) [inline library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: This will be prone to memory leaks where the map is closed. 
I think tracking if location component was initialized should be checked inside the KujakuMap view context via a flag and not as a static variable.

File: library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+    private MapboxLocationComponentWrapper() {}
+
+    @SuppressWarnings( {"MissingPermission"})
+    public static void init(@NonNull MapboxMap mapboxMap,@NonNull Context context) {
+        mapboxLocationComponentWrapper = new MapboxLocationComponentWrapper();
+        LocationComponent locationComponent = mapboxMap.getLocationComponent();
+        locationComponent.activateLocationComponent(context, (LocationEngine) null);
+        locationComponent.setLocationComponentEnabled(true);
+        locationComponent.setCameraMode(CameraMode.NONE);
```
Comment (githengi) [inline library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: Consider making the camera mode configurable e.g allowing definition in the layout xml

File: library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        locationComponent.setRenderMode(RenderMode.NORMAL);
```
Comment (githengi) [inline library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: Consider making the camera mode configurable e.g allowing definition in the layout xml

Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: You can always retrieve the LocationComponent from the LocationComponentWrapper and change any of these settings programatically.

File: library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java (hunk 8 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: Same as below.

File: library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java (hunk 6 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: The idea behind the class is that if in future we need more control over the LocationComponent, this will be the place to enforce those policies.

KujakuMapView is already large and this avoids adding unnecessary bulk to it.

File: library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java (hunk 7 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: I don't think I understand this concern about memory leakage. Could you rephrase?

Secondly, and as mentioned above, we should be trying to move functionality away from KujakuMapView?

File: library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java (hunk 5 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: Any reason for this?

File: library/build.gradle (hunk 1 above)
Comment (vincent-karuri) [reply library/build.gradle]: I need it to be a higher version?  

I want to publish it. What is you're suggestion?

File: library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java (hunk 3 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: @githengi  could you test this part and provide feedback?

File: library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java (hunk 5 above)
Comment (githengi) [reply library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: the internal created a variable that does nothing, you just assign it to the global variable

File: library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java (hunk 7 above)
Comment (githengi) [reply library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: when mapview is closed it will not be reclaimed by gc because of this static variable which will still hold a reference to the mapview.

File: library/build.gradle (hunk 1 above)
Comment (ekigamba) [reply library/build.gradle]: Kindly create a different PR when making a release when you are done with the feature(s) required for that release and merged to master

File: library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java (hunk 7 above)
Comment (githengi) [reply library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: The better design would be define `MapboxLocationComponentWrapper` as an a field in the KujakuMapView

File: sample/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -63,5 +64,7 @@
 
     <!-- WMTS Capabilities Url -->
     <string name="wmts_capabilities_url">https://tpwd.texas.gov/arcgis/rest/services/Vegetation_Mapping/Texas_Ecological_Mapping_Systems_Data/mapserver/WMTS/1.0.0/WMTSCapabilities.xml</string>
+    <string name="title_activity_configurable_location_circle">ConfigurableLocationCircleActivity
```
Comment (ekigamba) [inline sample/src/main/res/values/strings.xml]: Currently this is not functional. It is not used in the actual title rendering and thus not required

File: library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java (hunk 3 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: It's jerky because the `centerMap` function has a default 1000ms before settling, if I reduced it to 50ms, it is okay.

Any reason it was set to 1000ms?

File: library/build.gradle (hunk 1 above)
Comment (vincent-karuri) [reply library/build.gradle]: Cool thanks.

File: library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java (hunk 3 above)
Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: That's the animation duration. Which works OK in the master branch? It stutters while trying to change target location and zoom which if it takes too long(I assume it just did not zoom completely and re-click the button)

File: library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java (hunk 4 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: Cool.

File: library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java (hunk 3 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: I’m trying to find out if the duration can be reduced

it’s the one causing the jerky movement

Comment (githengi) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: The transition to the user location takes time sometimes. Also the transition is not smooth.

However on clicking the increase or decrease radius the response is instant and transition it better

File: library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java (hunk 7 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: Fixed.

File: library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java (hunk 5 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: Removed.

File: library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java (hunk 4 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/helpers/MapboxLocationComponentWrapper.java]: Fixed.

File: sample/src/main/res/values/strings.xml (hunk 10 above)
Comment (vincent-karuri) [reply sample/src/main/res/values/strings.xml]: Done.

File: library/src/main/java/io/ona/kujaku/exceptions/LocationComponentInitializationException.java (hunk 2 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/exceptions/LocationComponentInitializationException.java]: This has since been removed.

File: library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java (hunk 3 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: I have changed the transition time to 50ms.

I haven't find any issues with this so far.

Comment (githengi) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: This is not a great experience, where the user location just pops up. I think you should return the transition to 1 second.

Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: `setWarmGps(true);` in the onClick method of `currentLocationBtn` in `KujakuMapView` is what is causing jerky updates.

When disabled, the 1 second transitions are smooth.

I really can't tell what it does right now but it is causing some kind of loop that keeps updating locations while the animation is happening.

@ekigamba could you shed some light on what it does and maybe suggest a possible fix?

@githengi @ekigamba I again comment that `KujakuMapView` is growing too big to debug effectively and may need to be broken up.

Comment (githengi) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: @vincent-karuri I concur, The refactor should be prioritized, probably on the roadmap for the June release. @craigappl

Comment (githengi) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: @vincent-karuri @Kigamba maybe this is due to this call that is called on `setWarmGps(true)` `io.ona.kujaku.utils.LocationSettingsHelper#checkLocationEnabled`

Its making a location request independently

Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: 1. It was and is working OK on master (Where there are renders(similar as now) to create the same circle(s))
2. I have checked the code and there is no loop unless you can point it out
3. The issue has come up after adding this code so can you be certain that you cannot fix within the added code

Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: I don't think that's not the point.

I was asking what `setWarmGps` does and why it is called when the `currentLocationBtn` is clicked.

I know it works in master and `setWarmGps` calls this function `checkLocationSettingsAndStartLocationServices` which has so much going on (can't comprehend it).

The `checkLocationSettingsAndStartLocationServices` function is convoluted and not easy to understand, hence why I'm asking.

The point is `setWarmGps` is what is not playing nice with `LocationComponent` and that's why I'm asking you to provide more information on what it does.

Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: @vincent-karuri 

1. You moved this discussion to Slack yesterday and are yet to respond to the responses I provided let alone try out the solution that I provided since it seems the new code introduced is heavy on rendering. 
2. You went ahead and removed the transition that helps with the UX and ignored it - I hoped that this solution would help at least

Could you work sequentially on the actual problem? And try take the solutions provided. Otherwise, you are taking us in a circle if you can't specify some of these many issues(which I have checked) and do not want to try out the proposed solution. 

I honestly don't think it's fair or worth it to continue any argument and left it at a civil point

Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: FYI: I tried your solution out today. Didn't work. That's not the problem.

 I later discovered it had to do with the call to `setWarmGps` after doing my own debugging which took a lot of time as well.

That's when I sought to get your opinion on what might be happening in this function that could be a problem (I believe you were the core implementer).

This is allowed and even encouraged in software engineering. I'm yet to get a response trying to walk me through what may be going on in that function or the functions it calls.

I also specified exactly what function I'm unable to follow. I fail to see what the issue is here.

Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: > @vincent-karuri @Kigamba maybe this is due to this call that is called on `setWarmGps(true)` `io.ona.kujaku.utils.LocationSettingsHelper#checkLocationEnabled`
> 
> Its making a location request independently

Thanks for this input Sam. It was invaluable in tracking down the issue.

Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: @githengi @Kigamba take a look at the current implementation that I believe addresses the camera focus issue.

Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: > 1. It's a bit jerky. I don't know why?
> 2. The buffer radius and it’s blue appearance(of the location component) disappears after some time, is this going to be a problem for the specific use cases of Reveal and other host apps
>    screenshot_20190220-115336

The grey icon appears in a [stale state](https://github.com/mapbox/mapbox-android-demo/issues/735), clicking on the `current location button` fixes it.

I believe it is important to keep it, thoughts @githengi?

Comment (githengi) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: @vincent-karuri Great work. 
The location widget now transitions to users location with animation as expected.

Did you figure why after sometime the my location widget turns gray

Comment (githengi) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: @vincent-karuri we can keep it, however the 30 seconds default to turn my location to gray maybe too small. Change Default to maybe 2 minutes.

Comment (githengi) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: It would be nice if its configurable just like on mapbox, the host app can decide to disable it or provide the duration when the location becomes stale.

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java [hunk 11]
```diff
@@ -1169,9 +1146,10 @@ public void setWarmGps(boolean warmGps, @Nullable String rejectionDialogTitle, @
             hasAlreadyRequestedEnableLocation = false;
             // In case the location settings were turned off while the warmGps is still true, it means that the LocationClient is also on
             // We should just re-enable the location so that the LocationClient can reconnect to the location services
-            checkLocationSettingsAndStartLocationServices(shouldStartNow);
+            checkLocationSettingsAndStartLocationServices(shouldStartNow, onLocationServicesEnabledCallBack);
+        } else if (onLocationServicesEnabledCallBack != null) {
+            onLocationServicesEnabledCallBack.onSuccess();
```
Comment (githengi) [inline library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: Should this be called here?

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java [hunk 12]
```diff
@@ -208,11 +201,14 @@ private void init(@Nullable AttributeSet attributeSet) {
...
-
-                // Enable asking for enabling the location by resetting this flag in case it was true
-                hasAlreadyRequestedEnableLocation = false;
-                setWarmGps(true);
+                    // Enable asking for enabling the location by resetting this flag in case it was true
+                    hasAlreadyRequestedEnableLocation = false;
+                    setWarmGps(true, null, null, new OnLocationServicesEnabledCallBack() {
+                        @Override
+                        public void onSuccess() {
+                            focusOnUserLocation(true, locationBufferRadius);
```
Comment (githengi) [inline library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: This does handle cases where location were disabled and user accepted to enable them.

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java [hunk 13]
```diff
@@ -1059,7 +1026,7 @@ public void onResume() {
         getMapboxMap();
         // This prevents an overlay issue the first time when requesting for permissions
         if (warmGps && Permissions.check(getContext(), Manifest.permission.ACCESS_FINE_LOCATION)) {
-            checkLocationSettingsAndStartLocationServices(true);
+            checkLocationSettingsAndStartLocationServices(true, null);
```
Comment (githengi) [inline library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: This does handle cases where location were disabled and user accepted to enable them.

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 11 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: Yes, because location services are already enabled.

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 12 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: Sorry, I don't understand what this means, could you rephrase?

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 13 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: Sorry, I don't understand what this means, could you rephrase?

File: library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java (hunk 3 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/interfaces/IKujakuMapView.java]: All this is configurable via xml and `LocationComponentOptions`.

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 13 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: This does not have any call back so does this matter anyway?

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 12 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: @githengi how do you suggest this is handled?

`status.startResolutionForResult(activity, Constants.RequestCode.LOCATION_SETTINGS);` this is what checks the settings and returns it's result in an activity.

I guess a bigger question is whether clicking the button again after enabling significantly affect UX?

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 11 above)
Comment (githengi) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: This is called if `warmGps` is not true. Isn't that a bug

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 12 above)
Comment (githengi) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: Yes that's is not a  good UX, having to click my location again, users may not be able to easily know that they have to click it again.

Why not call `focusOnUserLocation(true, locationBufferRadius);` before or `setWarmGps` so that `onResume` takes user to their location.

This was the logic before as on master

Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: The whole point behind the callback approach is not to call `focusOnUserLocation(true, locationBufferRadius)` before.

I have added some logic in onResume that should help with this.

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 13 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: Same as above.

File: library/src/main/java/io/ona/kujaku/views/KujakuMapView.java (hunk 11 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/views/KujakuMapView.java]: True, I thought the check was for `shouldStartNow`.

File: sample/src/main/java/io/ona/kujaku/sample/activities/ConfigurableLocationCircleActivity.java [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+            public void onClick(View v) {
+                kujakuMapView.focusOnUserLocation(true, 70f);
+            }
+        });
+
+        Button btnDecreaseCircleRadius = findViewById(R.id.btn_decrease_circle_radius);
+        btnDecreaseCircleRadius.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                kujakuMapView.focusOnUserLocation(true, 10f);
```
Comment (githengi) [inline sample/src/main/java/io/ona/kujaku/sample/activities/ConfigurableLocationCircleActivity.java]: Call `io.ona.kujaku.views.KujakuMapView#setWarmGps(boolean, java.lang.String, java.lang.String, io.ona.kujaku.callbacks.OnLocationServicesEnabledCallBack) ` with on OnSuccess call `kujakuMapView.focusOnUserLocation(true, 10f);`

File: sample/src/main/java/io/ona/kujaku/sample/activities/ConfigurableLocationCircleActivity.java [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+        Mapbox.getInstance(this, BuildConfig.MAPBOX_SDK_ACCESS_TOKEN);
+
+        kujakuMapView = findViewById(R.id.configurable_circle_activity_map_view);
+        kujakuMapView.showCurrentLocationBtn(true);
+
+        Button btnIncreaseCircleRadius = findViewById(R.id.btn_increase_circle_radius);
+        btnIncreaseCircleRadius.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                kujakuMapView.focusOnUserLocation(true, 70f);
```
Comment (githengi) [inline sample/src/main/java/io/ona/kujaku/sample/activities/ConfigurableLocationCircleActivity.java]: Call `io.ona.kujaku.views.KujakuMapView#setWarmGps(boolean, java.lang.String, java.lang.String, io.ona.kujaku.callbacks.OnLocationServicesEnabledCallBack) ` with on OnSuccess call `kujakuMapView.focusOnUserLocation(true, 70f);`

Comment (vincent-karuri) [reply sample/src/main/java/io/ona/kujaku/sample/activities/ConfigurableLocationCircleActivity.java]: Done.

File: sample/src/main/java/io/ona/kujaku/sample/activities/ConfigurableLocationCircleActivity.java (hunk 14 above)
Comment (vincent-karuri) [reply sample/src/main/java/io/ona/kujaku/sample/activities/ConfigurableLocationCircleActivity.java]: Done.
